### PR TITLE
Add NDCG metric to rec_sys

### LINKED ISSUE #3581: [Feature]: Add NDCG (Normalized Discounted Cumulative Gain) metric to rec_sys
## 🚀 Feature

(Follow up to #3569 and #2631)

After adding `HitRate@top_k` and `MRR@top_k` metrics for recommendation systems, it makes sense to add `NDCG@top_k` as well.

NDCG complements MRR by considering all relevant items in the ranking, not just the first one. This makes it useful for:
- Recommendation systems evaluation
- RAG (Retrieval Augmented Generation) pipeline evaluation
- Information retrieval benchmarking

Similar to MRR, the implementation would:
- Follow the same structure as `HitRate` and `MRR`
- Support multiple `top_k` values in a single call
- Use ranx as reference implementation for tests

### PR DESCRIPTION
**Pull request by steaphenai**
Fixes #3581 
Refs #3568 #3569 #3610 

Description: Implements NDCG (Normalized Discounted Cumulative Gain) metric for recommendation systems and ranking evaluation.

## Implementation Details
- Follows the same pattern as `HitRate` and `MRR` metrics
- Uses `ranx` library for verification (addressing #3569 feedback)
-  Supports both binary (0/1) and graded (0-4) labels
- `relevance_threshold` defaults to 1.0 for binary labels
- Uses `stable=True` for reproducible tie-breaking
- Handles k > num_items
- Will align with top_k API changes from #3568 when finalized

## Mathematical Background
- **DCG**: Discounted Cumulative Gain - rewards relevant items, penalizes low positions
- **IDCG**: Ideal DCG - perfect ranking baseline
- **NDCG**: DCG / IDCG - normalized score from 0.0 to 1.0

Formula: `NDCG@K = Σ(2^rel_i - 1) / log2(i + 1)`

## Testing
- 26 tests passing locally (verified against ranx)
- Parametrized tests across multiple configurations
- Edge cases covered (perfect prediction, zero relevance, graded labels)
- Distributed training testedcs

Check list:
- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

**Comment by aaishwarymishra** [comment]
@steaphenai  In `docs/source/metrics.rst` add  `rec_sys.NDCG` in complete list of metrics to pass initial checks.

**Comment by steaphenai** [comment]
> @steaphenai In `docs/source/metrics.rst` add `rec_sys.NDCG` in complete list of metrics to pass initial checks.

@aaishwarymishra Thanks!!

File: tests/ignite/metrics/rec_sys/test_ndcg.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+
+def ranx_ndcg(
+    y_pred: np.ndarray,
+    y: np.ndarray,
+    top_k: list[int],
+    ignore_zero_hits: bool = True,
+) -> list[float]:
+    """Reference NDCG implementation using ranx for verification. https://github.com/AmenRa/ranx """
+    from ranx import Qrels, Run, evaluate
```
Comment (vfdev-5) [inline tests/ignite/metrics/rec_sys/test_ndcg.py]: this package is never installed in the CI?

Comment (steaphenai) [reply tests/ignite/metrics/rec_sys/test_ndcg.py]: @vfdev-5  Yes, exactly - ranx is not installed in the CI environment. I've fixed this now.
Used `try/except ImportError` to handle missing ranx  (ranx is set to None if not installed)
Wrapped ranx validation checks with `if ranx is not None:`
Added `pytest.skip()` for tests that specifically require ranx for validation
Thanks for catching this!

Comment (vfdev-5) [reply tests/ignite/metrics/rec_sys/test_ndcg.py]: @steaphenai I should install it in the CI and test the implementation on each run in the CI. Put ranx into requirements-dev.txt instead.

**Comment by vfdev-5** [comment]
@steaphenai I was checking your implementation compared to previous ones where contributors were blocked with vectorized computation of 
```python
discounted_gains = torch.tensor(
  [_tie_averaged_dcg(y_p, y_t, discount_cumsum, device) for y_p, y_t in zip(y_pred, y_true)], device=device
)
```
when ignore_ties is False. I can't find the same thing in your implementation, can you explain your implementation, what it based on?

**Comment by steaphenai** [comment]
> @steaphenai I was checking your implementation compared to previous ones where contributors were blocked with vectorized computation of
> 
> ```python
> discounted_gains = torch.tensor(
>   [_tie_averaged_dcg(y_p, y_t, discount_cumsum, device) for y_p, y_t in zip(y_pred, y_true)], device=device
> )
> ```
> 
> when ignore_ties is False. I can't find the same thing in your implementation, can you explain your implementation, what it based on?

@vfdev-5 My implementation is based on the standard NDCG formula and validated against the ranx library.
For tie handling, I use `torch.argsort(..., stable=True)` which breaks ties deterministically by index order (same as MRR). This ensures reproducible results and matches ranx's behavior.
I validated correctness by testing against ranx across various scenarios, all tests pass with matching results.

**Comment by steaphenai** [comment]
@vfdev-5 Just to clarify on the vectorization part, my implementation is fully vectorized using PyTorch operations (no Python loops).

Instead of looping through each example like:
```python
[_tie_averaged_dcg(y_p, y_t, ...) for y_p, y_t in zip(y_pred, y_true)]
```

I use batched PyTorch operations:
```python
ranked_indices = torch.argsort(y_pred, dim=-1, descending=True, stable=True)[:, :max_k]
ranked_relevance = torch.gather(y_for_dcg, dim=-1, index=ranked_indices)
```

This processes all examples in the batch simultaneously, which should be more efficient than the loop-based approach.

**Comment by vfdev-5** [comment]
To help with reviewing can you please run catalyst and ranx on some non-trivial examples involving ties etc and compare results. 
- https://github.com/catalyst-team/catalyst/blob/master/catalyst/metrics/_ndcg.py
- test cases: https://github.com/catalyst-team/catalyst/blob/master/tests/catalyst/metrics/functional/test_ndcg.py

File: ignite/metrics/rec_sys/ndcg.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+    .. math::
+        \text{DCG}@K = \sum_{i=1}^{K} \frac{2^{\text{rel}_i} - 1}{\log_2(i + 1)}
+
+    and :math:`\text{rel}_i` is the relevance score of the item at position :math:`i` in the
+    ranked list (1-indexed).
+
+    - ``update`` must receive output of the form ``(y_pred, y)``.
+    - ``y_pred`` is expected to be raw logits or probability score for each item in the catalog.
+    - ``y`` is expected to contain relevance scores (can be binary or graded).
```
Comment (vfdev-5) [inline ignite/metrics/rec_sys/ndcg.py:40]: what means "graded"?

File: ignite/metrics/rec_sys/ndcg.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            metric = NDCG(top_k=[1, 2, 3, 4])
+            metric.attach(default_evaluator, "ndcg")
+            y_pred=torch.Tensor([
+                [4.0, 2.0, 3.0, 1.0],
+                [1.0, 2.0, 3.0, 4.0]
+            ])
+            y_true=torch.Tensor([
+                [0.0, 0.0, 1.0, 1.0],
+                [0.0, 0.0, 0.0, 0.0]
+            ])
```
Comment (vfdev-5) [inline ignite/metrics/rec_sys/ndcg.py:108]: is it appropriate to have y_true as float?

File: ignite/metrics/rec_sys/ndcg.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        if self.ignore_zero_hits:
+            valid_mask = torch.any(y >= self.relevance_threshold, dim=-1)
+            y_pred = y_pred[valid_mask]
+            y = y[valid_mask]
+
+        if y.shape[0] == 0:
+            return
+
+        # Zero out items below relevance threshold for DCG computation
+        y_for_dcg = torch.where(y >= self.relevance_threshold, y, torch.zeros_like(y))
```
Comment (vfdev-5) [inline ignite/metrics/rec_sys/ndcg.py]:
```suggestion
        y_for_dcg = torch.where(y >= self.relevance_threshold, y, 0)
```
?

File: ignite/metrics/rec_sys/ndcg.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+        max_k = self.top_k[-1]
+
+        # Get ranked indices based on predictions (stable=True for deterministic tie-breaking)
+        ranked_indices = torch.argsort(y_pred, dim=-1, descending=True, stable=True)[:, :max_k]
+        
+        # Get relevance scores in the predicted ranking order
+        ranked_relevance = torch.gather(y_for_dcg, dim=-1, index=ranked_indices)
```
Comment (vfdev-5) [inline ignite/metrics/rec_sys/ndcg.py]: we can refactor these two lines as catalyst does: https://github.com/catalyst-team/catalyst/blob/e99f90655d0efcf22559a46e928f0f98c9807ebf/catalyst/metrics/functional/_misc.py#L77-L102

File: ignite/metrics/rec_sys/ndcg.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+        # Compute ideal ranking by sorting true relevance scores
+        ideal_relevance = torch.sort(y_for_dcg, dim=-1, descending=True, stable=True)[0][:, :max_k]
```
Comment (vfdev-5) [inline ignite/metrics/rec_sys/ndcg.py:212]: I'm not sure where this comes from and why we should sort targets?

File: ignite/metrics/rec_sys/ndcg.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+            [0.0, 0.31..., 0.31..., 0.31...]
+
+    .. versionadded:: 0.6.0
+    """
+
+    required_output_keys = ("y_pred", "y")
+    _state_dict_all_req_keys = ("_sum_ndcg_per_k", "_num_examples")
+
+    def __init__(
```
Comment (vfdev-5) [inline ignite/metrics/rec_sys/ndcg.py:143]: Should we also support two types of gain_function: exp_rank and linear_rank, https://github.com/catalyst-team/catalyst/blob/master/catalyst/metrics/functional/_ndcg.py#L9

File: ignite/metrics/rec_sys/ndcg.py (hunk 2 above)
Comment (steaphenai) [reply ignite/metrics/rec_sys/ndcg.py:40]: @vfdev-5 "Graded" means the relevance labels can have multiple levels beyond just binary (0/1).
Like for example, Binary relevance: 0 (not relevant) or 1 (relevant)
Graded relevance: 0 (not relevant), 1 (marginally relevant), 2 (relevant), 3 (highly relevant), 4 (perfectly relevant). The NDCG formula naturally handles graded relevance through the gain function: 2^rel - 1.
The `relevance_threshold` parameter lets users control which grades count as "relevant" for filtering.

File: ignite/metrics/rec_sys/ndcg.py (hunk 3 above)
Comment (steaphenai) [reply ignite/metrics/rec_sys/ndcg.py:108]: Yes, y_true should be float to support graded relevance scores (e.g., 0.0, 1.0, 2.0, 3.0, 4.0). Binary labels work fine as floats too (0.0, 1.0). This is consistent with how ranx and other IR metrics handle relevance labels.

File: ignite/metrics/rec_sys/ndcg.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/metrics/rec_sys/ndcg.py:40]: Put this explanation to the docs please

File: ignite/metrics/rec_sys/ndcg.py (hunk 5 above)
Comment (steaphenai) [reply ignite/metrics/rec_sys/ndcg.py]: @vfdev-5 I looked at the Catalyst code you linked. My implementation already uses the same approach:

Catalyst does:
```python
outputs_order = torch.argsort(outputs, descending=True, dim=-1)
targets_sorted = torch.gather(targets, dim=-1, index=outputs_order)
```
In our implementation:
```python
ranked_indices = torch.argsort(y_pred, dim=-1, descending=True, stable=True)[:, :max_k]
ranked_relevance = torch.gather(y_for_dcg, dim=-1, index=ranked_indices)
```
The only differences are:
we added `stable=True` for deterministic tie-breaking and we slice to `[:, :max_k]` since we only need top-k items

Does this approach look good? or do we need to make a change?

File: ignite/metrics/rec_sys/ndcg.py (hunk 6 above)
Comment (steaphenai) [reply ignite/metrics/rec_sys/ndcg.py:212]: This computes the Ideal DCG (IDCG) - the best possible DCG if items were perfectly ranked by their true relevance.
NDCG formula: NDCG = DCG / IDCG
To get IDCG, we need to Sort items by their TRUE relevance scores (descending) and Compute DCG on this perfect ranking.

This gives us the maximum possible DCG value, which we use to normalize the actual DCG. This is standard in the NDCG formula and we can see the same approach in ranx and other NDCG implementations.

File: ignite/metrics/rec_sys/ndcg.py (hunk 5 above)
Comment (vfdev-5) [reply ignite/metrics/rec_sys/ndcg.py]: I mean put these two lines into a helper function such that other recsys metrics could use it

File: ignite/metrics/rec_sys/ndcg.py (hunk 6 above)
Comment (vfdev-5) [reply ignite/metrics/rec_sys/ndcg.py:212]: > This is standard in the NDCG formula and we can see the same approach in ranx and other NDCG implementations.

Please provide links to this op in catalyst and ranx

File: ignite/metrics/rec_sys/ndcg.py (hunk 7 above)
Comment (steaphenai) [reply ignite/metrics/rec_sys/ndcg.py:143]: @vfdev-5 I've Added support for both gain_function types (exp_rank and linear_rank) as you suggested!

**Comment by steaphenai** [comment]
@vfdev-5 I added `ignite.metrics.rec_sys.NDCG` with support for `top_k`, `ignore_zero_hits`, and `relevance_threshold`. I also added a shared `_get_ranked_items(...)` helper for the ranking logic and added `gain_function` support (`exp_rank` and `linear_rank`). The NDCG docstring and docs were updated to clarify graded relevance and for validation, I added tests in tests/ignite/metrics/rec_sys/test_ndcg.py which parity checks against `ranx` and `Catalyst`, plus tie cases and edge cases (perfect ranking, zero relevance, threshold behavior, multi-batch accumulation, invalid inputs, and distributed integration).

File: tests/ignite/metrics/rec_sys/test_ndcg.py [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+    y_pred: np.ndarray,
+    y: np.ndarray,
+    top_k: list[int],
+    ignore_zero_hits: bool = True,
+    relevance_threshold: float = 1.0,
+    gain_function: str = "exp_rank",
+) -> list[float]:
+    """Reference NDCG implementation using catalyst for verification."""
+    pytest.importorskip("catalyst", reason="catalyst is required for catalyst parity checks")
+    from catalyst.metrics.functional import ndcg as catalyst_ndcg_fn
```
Comment (vfdev-5) [inline tests/ignite/metrics/rec_sys/test_ndcg.py]: I prefer we either install catalyst and check against it or register the expected output for a number of test cases and check against them.
Does ranx have the same option as catalyst and can compute the same results under the same options?

File: ignite/metrics/rec_sys/ndcg.py (hunk 6 above)
Comment (vfdev-5) [reply ignite/metrics/rec_sys/ndcg.py:212]: @steaphenai check above comment to provide the link on sorting op

File: tests/ignite/metrics/rec_sys/test_ndcg.py [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+
+@pytest.mark.parametrize("num_queries", [1, 10, 100])
+@pytest.mark.parametrize("num_items", [5, 20, 100])
+@pytest.mark.parametrize("k", [1, 5, 10])
+@pytest.mark.parametrize("ignore_zero_hits", [True, False])
+def test_compute_vs_ranx(num_queries, num_items, k, ignore_zero_hits, available_device):
+    """Verify NDCG matches ranx across a wide range of input shapes and k values."""
+    torch.manual_seed(42)
+    y_pred = torch.randn(num_queries, num_items)
+    y_true = torch.randint(0, 2, (num_queries, num_items)).float()
```
Comment (vfdev-5) [inline tests/ignite/metrics/rec_sys/test_ndcg.py]:
```suggestion
    y_true = torch.randint(0, 5, (num_queries, num_items)).float()
```

**Comment by vfdev-5** [comment]
@steaphenai why do you close this PR ?

**Comment by steaphenai** [comment]
> @steaphenai why do you close this PR ?

@vfdev-5 I'm extremely sorry, I unfortunately deleted the local branch

**Comment by vfdev-5** [comment]
@steaphenai you have the copy on GH and locally, you can restore the deleted one from either source.
Let's close the new PR and continue the dev here as it was already reviewed and contain comments.
Next time, please first ask on discord for help, maybe this can help avoid extra effort.

File: ignite/metrics/rec_sys/ndcg.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+        top_k: list[int],
+        ignore_zero_hits: bool = True,
+        relevance_threshold: float = 1.0,
+        gain_function: str = "exp_rank",
+        output_transform: Callable = lambda x: x,
+        device: str | torch.device = torch.device("cpu"),
+        skip_unrolling: bool = False,
+    ):
+        if any(k <= 0 for k in top_k):
+            raise ValueError(" top_k must be list of positive integers only.")
```
Comment (vfdev-5) [inline ignite/metrics/rec_sys/ndcg.py]:
```suggestion
            raise ValueError(f"top_k must be list of positive integers only, but given {top_k}")
```

File: ignite/metrics/rec_sys/ndcg.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+
+        if gain_function not in ["exp_rank", "linear_rank"]:
+            raise ValueError("gain_function must be either 'exp_rank' or 'linear_rank'")
```
Comment (vfdev-5) [inline ignite/metrics/rec_sys/ndcg.py]:
```suggestion
            raise ValueError(f"gain_function must be either 'exp_rank' or 'linear_rank', but given {gain_function}")
```

**Comment by vfdev-5** [comment]
@steaphenai please check the failing CI job: https://github.com/pytorch/ignite/actions/runs/23947698404/job/70144090725?pr=3608

**Comment by steaphenai** [comment]
> @steaphenai please check the failing CI job: https://github.com/pytorch/ignite/actions/runs/23947698404/job/70144090725?pr=3608

`PS C:\ignite> pre-commit run --all-files
nb-clean.................................................................Passed
check toml...............................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
fix end of files.........................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
prettier.................................................................Passed
ruff check...............................................................Passed
ruff format..............................................................Passed`

@vfdev-5 I verified with pre-commit run --all-files.  All the tests were passed locally.

File: tests/ignite/metrics/rec_sys/test_ndcg.py [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+def ranx_ndcg(
+    y_pred: np.ndarray,
+    y: np.ndarray,
+    top_k: list[int],
+    ignore_zero_hits: bool = True,
+    relevance_threshold: float = 1.0,
+    gain_function: str = "exp_rank",
+) -> list[float]:
+    """Reference NDCG implementation using ranx for verification. https://github.com/AmenRa/ranx"""
+    from ranx import Qrels, Run, evaluate
```
Comment (vfdev-5) [inline tests/ignite/metrics/rec_sys/test_ndcg.py]: Let's put this import to the top of the file.

File: tests/ignite/metrics/rec_sys/test_ndcg.py [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+    y_pred: np.ndarray,
+    y: np.ndarray,
+    top_k: list[int],
+    ignore_zero_hits: bool = True,
+    relevance_threshold: float = 1.0,
+    gain_function: str = "exp_rank",
+) -> list[float]:
+    """Reference NDCG implementation using catalyst for verification."""
+    pytest.importorskip("catalyst", reason="catalyst is required for catalyst parity checks")
+    from catalyst.metrics.functional import ndcg as catalyst_ndcg_fn
```
Comment (vfdev-5) [inline tests/ignite/metrics/rec_sys/test_ndcg.py]: Same here, put the import to the top of the file

**Review by vfdev-5 — COMMENTED** [review]
Thanks for updates @steaphenai 
We almost there, just few updates

**Comment by steaphenai** [comment]
> Thanks for updates @steaphenai We almost there, just few updates

@vfdev-5 Done! I've set the imports to the top of the file.

**Review by vfdev-5 — APPROVED** [review]
Thanks a lot for the PR @steaphenai , LGTM

File: tests/ignite/metrics/rec_sys/test_ndcg.py [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+    y_pred: np.ndarray,
+    y: np.ndarray,
+    top_k: list[int],
+    ignore_zero_hits: bool = True,
+    relevance_threshold: float = 1.0,
+    gain_function: str = "exp_rank",
+) -> list[float]:
+    """Reference NDCG implementation using catalyst for verification."""
+    pytest.importorskip("catalyst", reason="catalyst is required for catalyst parity checks")
+
```
Comment (vfdev-5) [inline tests/ignite/metrics/rec_sys/test_ndcg.py]:
```suggestion
```

**Comment by vfdev-5** [comment]
@steaphenai still failing CI code style check, please fix it

**Comment by steaphenai** [comment]
@vfdev-5 I moved the Catalyst import to the top-level as requested, but TPU CI was failing during test collection because importing catalyst triggers a `torch_xla` linkage error in that environment. I am thinking it to update it to a guarded top-level import (try/except) and marked only the Catalyst-dependent tests with skip-if when Catalyst is unavailable. This keeps import at top while avoiding unrelated CI breakage. Is this approach okay for this PR?
Something like this
```
from ranx import Qrels, Run, evaluate
try:
    from catalyst.metrics.functional import ndcg as catalyst_ndcg_fn

    HAS_CATALYST = True
except Exception:  
    catalyst_ndcg_fn = None
    HAS_CATALYST = False
    ```

**Comment by vfdev-5** [comment]
Catalyst is unmaintained now, so I was actually thinking if we can remove it from the tests and keep only ranx (which is also look a bit left aside).
We can simply remove all the tests involving catalyst. Other option is to do what you suggested.
What do you think?

**Comment by steaphenai** [comment]
@vfdev-5 Makes sense. I agree we should remove Catalyst-based parity tests and keep ranx-based validation. I’ll drop Catalyst imports/helpers/tests and keep the ranx checks as the reference path
